maven-model/src/main/java/org/apache/maven/model/merge/ModelMerger.java

```diff
@@ -3018,4 +2618,298 @@ protected Object getExclusionKey( Exclusion exclusion )
+    {
+        @Override
+        public Object key( Profile profile )
+        {
+            return getProfileKey( profile );
+        }
+    }
+
+    class RepositoryKeyComputer implements KeyComputer<Repository>
+    {
+        @Override
+        public Object key( Repository repository )
+        {
+            return getRepositoryKey( repository );
+        }
+    }
+
+    class ReportPluginKeyComputer implements KeyComputer<ReportPlugin>
+    {
+        @Override
+        public Object key( ReportPlugin plugin )
+        {
+            return getReportPluginKey( plugin );
+        }
+    }
+
+    class PluginKeyComputer implements KeyComputer<Plugin>
+    {
+        @Override
+        public Object key( Plugin plugin )
+        {
+            return getPluginKey( plugin );
+        }
+    }
+
+    class ReportSetKeyComputer implements KeyComputer<ReportSet>
+    {
+        @Override
+        public Object key( ReportSet reportSet )
+        {
+            return getReportSetKey( reportSet );
+        }
+    }
+
+    class NotifierKeyComputer implements KeyComputer<Notifier>
+    {
+        @Override
+        public Object key( Notifier notifier )
+        {
+            return getNotifierKey( notifier );
+        }
+    }
+
+    class ExtensionKeyComputer implements KeyComputer<Extension>
+    {
+        @Override
+        public Object key( Extension extension )
+        {
+            return getExtensionKey( extension );
+        }
+    }
+
+    class ResourceKeyComputer implements KeyComputer<Resource>
+    {
+        @Override
+        public Object key( Resource resource )
+        {
+            return getResourceKey( resource );
+        }
+    }
+
+    class ExecutionKeyComputer implements KeyComputer<PluginExecution>
+    {
+        @Override
+        public Object key( PluginExecution pluginExecution )
+        {
+            return getPluginExecutionKey( pluginExecution );
+        }
+    }
+
+    class ExclusionKeyComputer implements KeyComputer<Exclusion>
+    {
+        @Override
+        public Object key( Exclusion exclusion )
+        {
+            return getExclusionKey( exclusion );
+        }
+    }
+
+    static <T> List<T> merge( List<T> tgt, List<T> src, boolean sourceDominant, KeyComputer<T> computer )
+    {
+        if ( src.isEmpty() )
+        {
+            return tgt;
+        }
+
+        MergingList<T> list;
+        if ( tgt instanceof MergingList )
+        {
+            list = (MergingList<T>) tgt;
+        }
+        else
+        {
+            list = new MergingList<>( computer, src.size() + tgt.size() );
+            list.mergeAll( tgt, true );
+        }
+
+        list.mergeAll( src, sourceDominant );
+        return list;
+    }
+
+    /**
+     * Merging list
+     * @param <V>
+     */
+    public static class MergingList<V> extends AbstractList<V>
+    {
+        private final KeyComputer<V> keyComputer;
+        private Map<Object, V> map;
+        private List<V> list;
```

Comment:
It is not clear why we have to deal with these two dual modes.
We start as a map and then we move to a list depending on the sequence of method calls.

I wonder if there is some way to make it simpler.

Are we always ending in list-based mode?
If so, why not explicitly having a map based stage then a list based stage, without the possibility to be back to map mode.

Reply:
If I see one field `final` and other two not final, I should ask you guys about thread safety. Is this object created in thread T1 and used in T2 or T3?

Reply:
`private Map<Object, V> map;`
Object has no guarante that equals and hashCode are implemented and so that the Map would work properly with equal objects.

Reply:
This is an internal data structure, there is no need to complicate the requirements over the values.
Do you think it won't work as expected?

Reply:
@Tibor17 this is an internal object, the keys are the same that were used in Maps previously, so this is similar to the current code.
Thread safety is not a concern because merging happens in a single thread, and the current code uses an ArrayList to hold objects, which is not thread safe either.
@eolivelli the point is to avoid switching back and forth between maps and list during the merging time.  

Ideally, the model would be changed so that Plugin, Dependency and other merged objects would implement `equals` and `hashCode` instead of relying on `computeXxxKey`.  And those objects would be stored in a `Set` rather than a `List`.  But I didn't want to break compatibility at all, so I haven't changed the data structures.
I also did not want to risk breaking compatibility by using a non complete List implementation, so I decided to revert to the List whenever it is actually accessed (which does not happen during project build, but could be in maven plugins later).
I've chosen to try to keep full compatibility, but if you prefer a more conside code without the duplicate map/list backend, I can refactor, but there will be a risk of breaking compatibility.

Reply:
@gnodet thank you for your answers
I am fine with your current approach, I only wanted to understand better your change.

Reply:
I think the `MergingList` class would be better as private, which may help removing some of the problems.

Reply:
yes, I agree

Reply:
@gnodet 
@eolivelli 
See the usages and inheritance. This class is inherited 6 other classes - injector etc.
The CPU does not care if we call `MergingList` an internal project class. It sees main shared memory, registers and CPU instructions for reconsiling these two memories. The CPU has totally different view. It can see rules for `synchronized`, `volatile`, `final` but not our internal terminology and most probably we can see different reality than the CPU.

There is one guy, whose name is Doug Lea as an author of Java Concurrency, who was invited in design of many APIs and specs which topic was different from multithreading. That was because the implementation had to be capable in Java Enterprise which is naturally multiple threads capable.
The Maven should have such guy and then Maven won't be slow, won't eat too much memory and the implementation will have certain guarantees in memory execution and safety .

Reply:
@Tibor17 again, your objections may be totally valid but my goal was to work with 100% compatibility, so only internal refactoring and improvements, not rewriting maven ;-)

Reply:
@gnodet 
I don't want to stop you. It was not my intention of course, and therefore I said that we can talk about global picture at the ASF conference in Berlin.

I have very conservative development approach.
Normally, the design of the s/w has to be initially made with respect to many criteria and it is difficult to meet them in the early beginning. And every rework goes with pain later and later. That's the life.